File: frontend/src/api/income.ts
import client from './client';

export interface Income {
  id: string;
  amount: number;
  currency: string;
  source: string;
  income_date: string;
  is_recurring: boolean;
  frequency?: string;
  notes?: string;
  created_at: string;
}

export interface CreateIncome {
  amount: number;
  currency?: string;
  source: string;
  income_date: string;
  is_recurring?: boolean;
  frequency?: string;
  notes?: string;
}

export const incomeApi = {
  getAll: () => client.get<Income[]>('/api/income/').then((r) => r.data),
  create: (data: CreateIncome) => client.post<Income>('/api/income/', data).then((r) => r.data),
  update: (id: string, data: Partial<CreateIncome>) =>
    client.patch<Income>(`/api/income/${id}`, data).then((r) => r.data),
  delete: (id: string) => client.delete(`/api/income/${id}`),
};

File: backend/app/models/income.py
import uuid
from datetime import datetime, date
from typing import Optional
from sqlalchemy import String, Float, Date, DateTime, Text, Boolean, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import UUID

from app.database import Base


class Income(Base):
    __tablename__ = "income"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[str] = mapped_column(String(128), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    amount: Mapped[float] = mapped_column(Float, nullable=False)
    currency: Mapped[str] = mapped_column(String(10), default="EUR")
    source: Mapped[str] = mapped_column(String(255), nullable=False)
    income_date: Mapped[date] = mapped_column(Date, nullable=False)
    is_recurring: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    frequency: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
